manifest/v1alpha/slo/validation_composite_test.go

```diff
@@ -0,0 +1,258 @@
+	t.Run("fails - weight is zero for second composite objective", func(t *testing.T) {
+		slo := validCompositeSLO()
+		slo.Spec.Objectives[0].Composite.Objectives[1].Weight = 0.0
+		err := validate(slo)
+
+		testutils.AssertContainsErrors(t, slo, err, 1, testutils.ExpectedError{
+			Prop:    "spec.objectives[0].composite.components[0].objectives[1].weight",
+			Code:    validation.ErrorCodeRequired,
+			Message: "property is required but was empty",
+		})
+	})
+	t.Run("fails - one of objectives is the composite SLO itself (cycle)", func(t *testing.T) {
+		slo := validCompositeSLO()
+		slo.Spec.Objectives[0].Composite.Objectives[0].Project = "composite-project"
+		slo.Spec.Objectives[0].Composite.Objectives[0].Objective = "my-composite-slo"
+		err := validate(slo)
+
+		testutils.AssertContainsErrors(t, slo, err, 1, testutils.ExpectedError{
+			Prop:    "slo",
+			Code:    validation.ErrorCodeForbidden,
+			Message: "composite SLO cannot have itself as one of its objectives",
+		})
+	})
+	t.Run("fails - invalid objective project name", func(t *testing.T) {
+		slo := validCompositeSLO()
+		slo.Spec.Objectives[0].Composite.Objectives[0].Project = "composite/project"
+		err := validate(slo)
+
+		testutils.AssertContainsErrors(t, slo, err, 1, testutils.ExpectedError{
+			Prop:    "spec.objectives[0].composite.components[0].objectives[0].project",
+			Code:    joinErrorCodes(validation.ErrorCodeStringIsDNSSubdomain, validation.ErrorCodeStringMatchRegexp),
+			Message: expectedStringIsDNSSubdomainError,
+		})
+	})
+	t.Run("fails - invalid objective slo name", func(t *testing.T) {
+		slo := validCompositeSLO()
+		slo.Spec.Objectives[0].Composite.Objectives[0].SLO = "my-slo/alpha"
+		err := validate(slo)
+
+		testutils.AssertContainsErrors(t, slo, err, 1, testutils.ExpectedError{
+			Prop:    "spec.objectives[0].composite.components[0].objectives[0].slo",
+			Code:    joinErrorCodes(validation.ErrorCodeStringIsDNSSubdomain, validation.ErrorCodeStringMatchRegexp),
+			Message: expectedStringIsDNSSubdomainError,
+		})
+	})
+	t.Run("fails - invalid underlying objective name", func(t *testing.T) {
+		slo := validCompositeSLO()
+		slo.Spec.Objectives[0].Composite.Objectives[0].Objective = "go/od"
+		err := validate(slo)
+
+		testutils.AssertContainsErrors(t, slo, err, 1, testutils.ExpectedError{
+			Prop:    "spec.objectives[0].composite.components[0].objectives[0].objective",
+			Code:    joinErrorCodes(validation.ErrorCodeStringIsDNSSubdomain, validation.ErrorCodeStringMatchRegexp),
+			Message: expectedStringIsDNSSubdomainError,
+		})
+	})
+	t.Run("fails - weight less than zero", func(t *testing.T) {
+		slo := validCompositeSLO()
+		slo.Spec.Objectives[0].Composite.Objectives[0].Weight = -1.0
+		err := validate(slo)
+
+		testutils.AssertContainsErrors(t, slo, err, 1, testutils.ExpectedError{
+			Prop:    "spec.objectives[0].composite.components[0].objectives[0].weight",
+			Code:    validation.ErrorCodeGreaterThan,
+			Message: "should be greater than '0'",
```

Comment:
Testing the message tests for the value used by the validator, so it is useful. If someone were to change the validator to `-0.1`, the code would still be the same, but the test would fail, as we're checking the message.